retworkx-core/src/planar/lr_planar.rs

```diff
@@ -0,0 +1,699 @@
+        Interval { inner: None }
+    }
+}
+
+impl<T> Interval<T> {
+    fn new(low: T, high: T) -> Self {
+        Interval {
+            inner: Some((low, high)),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.inner.is_none()
+    }
+
+    fn unwrap(self) -> (T, T) {
+        self.inner.unwrap()
+    }
+
+    fn low(&self) -> Option<&T> {
+        match self.inner {
+            Some((ref low, _)) => Some(low),
+            None => None,
+        }
+    }
+
+    fn high(&self) -> Option<&T> {
+        match self.inner {
+            Some((_, ref high)) => Some(high),
+            None => None,
+        }
+    }
+
+    fn as_ref(&mut self) -> Option<&(T, T)> {
+        self.inner.as_ref()
+    }
+
+    fn as_mut(&mut self) -> Option<&mut (T, T)> {
+        self.inner.as_mut()
+    }
+
+    fn as_mut_low(&mut self) -> Option<&mut T> {
+        match self.inner {
+            Some((ref mut low, _)) => Some(low),
+            None => None,
+        }
+    }
+}
+
+impl<T> Interval<(T, T)>
+where
+    T: Copy + Hash + Eq,
+{
+    /// Returns ``true`` if the interval conflicts with ``edge``.
+    fn conflict<G>(&self, lr_state: &LRState<G>, edge: Edge<G>) -> bool
+    where
+        G: GraphBase<NodeId = T>,
+    {
+        match self.inner {
+            Some((_, ref h)) => lr_state.lowpt.get(h) > lr_state.lowpt.get(&edge),
+            _ => false,
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, PartialOrd)]
+struct ConflictPair<T> {
+    left: Interval<T>,
+    right: Interval<T>,
+}
+
+impl<T> Default for ConflictPair<T> {
+    fn default() -> Self {
+        ConflictPair {
+            left: Interval::default(),
+            right: Interval::default(),
+        }
+    }
+}
+
+impl<T> ConflictPair<T> {
+    fn new(left: Interval<T>, right: Interval<T>) -> Self {
+        ConflictPair { left, right }
+    }
+
+    fn swap(&mut self) {
+        std::mem::swap(&mut self.left, &mut self.right)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.left.is_empty() && self.right.is_empty()
+    }
+}
+
+impl<T> ConflictPair<(T, T)>
+where
+    T: Copy + Hash + Eq,
+{
+    /// Returns the lowest low point of a conflict pair.
+    fn lowest<G>(&self, lr_state: &LRState<G>) -> usize
+    where
+        G: GraphBase<NodeId = T>,
+    {
+        match (self.left.low(), self.right.low()) {
+            (Some(l_low), Some(r_low)) => lr_state.lowpt[l_low].min(lr_state.lowpt[r_low]),
+            (Some(l_low), None) => lr_state.lowpt[l_low],
+            (None, Some(r_low)) => lr_state.lowpt[r_low],
+            (None, None) => std::usize::MAX,
+        }
+    }
+}
+
+enum Sign {
+    Plus,
+    Minus,
+}
+
+/// Similar to ``DfsEvent`` plus an extra event ``FinishEdge``
+/// that indicates that we have finished processing an edge.
+enum LRTestDfsEvent<N> {
+    Finish(N),
+    TreeEdge(N, N),
+    BackEdge(N, N),
+    FinishEdge(N, N),
+}
+
+// An error: graph is *not* planar.
+struct NonPlanar {}
+
+struct LRState<G: GraphBase>
```

Comment:
Is there a reason for the use of `GraphBase` and `GraphProp` throughout instead of `StableGraph`? And is there a Python interface to `is_planar`?

Reply:
Our goal in retworkx-core is to provide functionality that can work for any graph data structure in the external petgraph library that implement the required traits. See https://docs.rs/petgraph/0.6.0/petgraph/visit/index.html#graph-trait-implementations.

The python interface is here:
https://github.com/Qiskit/retworkx/blob/b6f4a9212afe909d943aaaceba2b5d6230c8b437/src/planar/mod.rs#L33-L37